stash.1
.\"t
.\" Automatically generated by Pandoc 2.9.2
.\"
.TH "" "" "" "" ""
.hy
.SH stash
.PP
\f[C]stash\f[R] is a command line program for storing text data in
encrypted form.
.PP
All user data is encrypted using AES
256 (https://en.wikipedia.org/wiki/Advanced_Encryption_Standard) cipher.
.PP
Data is hashed for indexing purposes using a good random salt +
SHA512 (https://en.wikipedia.org/wiki/SHA-2).
.PP
.PP
\f[B]Table of Contents\f[R]
.IP \[bu] 2
stash
.RS 2
.IP \[bu] 2
Rationale
.IP \[bu] 2
Installation
.RS 2
.IP \[bu] 2
Linux and Mac
.IP \[bu] 2
Others
.RE
.IP \[bu] 2
Getting Started
.IP \[bu] 2
Browsing
.RS 2
.IP \[bu] 2
Terminal user interface (default)
.IP \[bu] 2
As Org text
.IP \[bu] 2
As Markdown text
.RE
.IP \[bu] 2
Dumping contents to stdout
.RS 2
.IP \[bu] 2
As Org text (default)
.IP \[bu] 2
As Markdown text
.IP \[bu] 2
As JSON text
.RE
.IP \[bu] 2
Backing up data
.IP \[bu] 2
Importing text
.IP \[bu] 2
Babashka pod support
.IP \[bu] 2
Customization
.IP \[bu] 2
Shell Autocompletion
.RS 2
.IP \[bu] 2
bash
.IP \[bu] 2
zsh
.IP \[bu] 2
fish
.RE
.IP \[bu] 2
Credits
.RE
.SS Rationale
.PP
All of us need to store sensitive information.
At the very least, we need to keep passwords and keys to numerous online
services we use.
Storing them in plain-text is risky.
So is not keeping backups in multiple locations.
.PP
\f[C]stash\f[R] is an open-source tool to store text in encrypted form.
The encrypted data can be stored in untrusted locations without
compromising it, granted the encryption-key is unguessable.
.SS Installation
.SS Linux and Mac
.PP
Install:
.IP
.nf
\f[C]
brew install rorokimdim/brew/stash
\f[R]
.fi
.PP
Upgrade:
.IP
.nf
\f[C]
brew upgrade stash
\f[R]
.fi
.PP
Uninstall:
.IP
.nf
\f[C]
brew uninstall stash
\f[R]
.fi
.PP
Binaries are available at
releases (https://github.com/rorokimdim/stash/releases).
.SS Others
.PP
No pre-built binaries available at this time.
We will need to build from source using \f[C]stack install\f[R] or
\f[C]cabal install\f[R].
.PP
Install stack (https://docs.haskellstack.org/en/stable/README/), clone
this repository and run the following in repository directory.
.IP
.nf
\f[C]
stack install
\f[R]
.fi
.SS Getting Started
.PP
Create a stash file.
.IP
.nf
\f[C]
stash create {path-to-stash-file}
\f[R]
.fi
.PP
Replace \f[C]{path-to-stash-file}\f[R] with path to the stash file.
For example \f[C]\[ti]/Dropbox/secret.stash\f[R].
.PP
Stash will prompt for the encryption-key (password) to use for the file.
The key is not saved anywhere, but a salted hash (good random salt +
SHA512) of the key and the salt is saved.
.PP
When we run a stash command, we will be prompted for our encryption key.
It is checked against the hash stored during \f[C]stash create\f[R].
For the duration of the command, the encryption key will be used for
encrypting/decrypting data.
.PP
The salt stored during \f[C]stash create\f[R] is also used for hashing
any data for indexing purposes.
.SS Browsing
.PP
This feature does not currently work on windows.
.SS Terminal user interface (default)
.IP
.nf
\f[C]
stash browse {path-to-stash-file}
\f[R]
.fi
.PP
.TS
tab(@);
l l.
T{
Shortcuts
T}@T{
Description
T}
_
T{
ESC (q)
T}@T{
Quit
T}
T{
?
T}@T{
Show this help
T}
T{
+
T}@T{
Add a new key
T}
T{
>
T}@T{
Add a child key to selected key
T}
T{
-
T}@T{
Delete selected key
T}
T{
, (r)
T}@T{
Rename selected key
T}
T{
/
T}@T{
Search and sort by pattern
T}
T{
y
T}@T{
Copy value of selected key into system clipboard
T}
T{
Enter
T}@T{
Set value of selected key
T}
T{
H
T}@T{
See history of values of selected key
T}
T{
Left arrow (h)
T}@T{
Move to parent of selected key
T}
T{
Right arrow (l)
T}@T{
Move to child of selected key
T}
T{
Up arrow (k, Ctrl-p)
T}@T{
Select above
T}
T{
Down arrow (j, Ctrl-n)
T}@T{
Select below
T}
T{
g, G
T}@T{
Move to top, Move to bottom
T}
T{
Ctrl-u, Ctrl-d
T}@T{
Scroll keys half page up / down
T}
T{
Ctrl-b, Ctrl-f
T}@T{
Scroll keys one page up / down
T}
.TE
.SS As Org text
.IP
.nf
\f[C]
stash browse -f org {path-to-stash-file}
\f[R]
.fi
.PP
Opens as Org text in default editor.
Set EDITOR environment variable to your editor command.
Defaults to vim.
.SS As Markdown text
.IP
.nf
\f[C]
stash browse -f markdown {path-to-stash-file}
\f[R]
.fi
.PP
Opens as markdown text in default editor.
Set EDITOR environment variable to your editor command.
Defaults to vim.
.SS Dumping contents to stdout
.SS As Org text (default)
.IP
.nf
\f[C]
stash dump -f org {path-to-stash-file}
\f[R]
.fi
.SS As Markdown text
.IP
.nf
\f[C]
stash dump -f markdown {path-to-stash-file}
\f[R]
.fi
.SS As JSON text
.IP
.nf
\f[C]
stash dump -f json {path-to-stash-file}
\f[R]
.fi
.SS Backing up data
.IP
.nf
\f[C]
stash backup {path-to-stash-file}
\f[R]
.fi
.PP
Creates a timestamped copy of stash file in the same directory as the
original file.
.SS Importing text
.PP
We can import \f[C]org\f[R] and \f[C]markdown\f[R] text into stash.
This feature is only available in version 0.2.0+.
.IP
.nf
\f[C]
cat {path-to-text-file} | stash import {path-to-stash-file}
\f[R]
.fi
.PP
\f[C]stash import\f[R] will assume text is in \f[C]org\f[R] format by
default, but \f[C]markdown\f[R] is supported too:
.IP
.nf
\f[C]
cat {path-to-text-file} | stash import -f markdown {path-to-stash-file}
\f[R]
.fi
.PP
We can import from another stash file similarly:
.IP
.nf
\f[C]
stash dump {path-to-source-stash-file} | stash import {path-to-destination-stash-file}
\f[R]
.fi
.PP
We will need to first enter encryption-key for the source stash-file
followed by encryption-key for the destination stash file.
.SS Babashka pod support
.PP
\f[C]stash\f[R] exposes a
bencode (https://en.wikipedia.org/wiki/Bencode) interface using Babashka
Pod Protocol (https://github.com/babashka/pods#the-protocol).
.PP
If you are familiar with clojure (https://clojure.org/) or
babashka (https://github.com/borkdude/babashka), see
clojure-example (https://github.com/rorokimdim/stash/tree/master/scripting-examples/clojure).
.PP
Common-lisp folks see
common-lisp-example (https://github.com/rorokimdim/stash/tree/master/scripting-examples/common-lisp).
.PP
Python folks see
python-example (https://github.com/rorokimdim/stash/tree/master/scripting-examples/python).
.SS Customization
.PP
A few things in stash can be customized via environment variables.
.PP
.TS
tab(@);
l l l l.
T{
Name
T}@T{
Description
T}@T{
Possible Values
T}@T{
Default
T}
_
T{
\f[C]STASH_ENCRYPTION_KEY\f[R]
T}@T{
encryption key
T}@T{
-
T}@T{
prompt as needed
T}
T{
\f[C]STASH_WIPE_CLIPBOARD_AFTER_BROWSE\f[R]
T}@T{
wipe clipboard after browse
T}@T{
true, false
T}@T{
false
T}
T{
\f[C]STASH_LOG_LEVEL\f[R]
T}@T{
logging level
T}@T{
DEBUG, INFO, WARN, ERROR
T}@T{
INFO
T}
T{
\f[C]STASH_TUI_COLOR_SELECTED\f[R]
T}@T{
color of selected item
T}@T{
see below
T}@T{
cyan
T}
T{
\f[C]STASH_TUI_COLOR_CURRENT_PATH\f[R]
T}@T{
color of current path
T}@T{
see below
T}@T{
white
T}
T{
\f[C]STASH_TUI_COLOR_SORT_PATTERN\f[R]
T}@T{
color of sort/search text
T}@T{
see below
T}@T{
white
T}
T{
\f[C]BABASHKA_POD\f[R]
T}@T{
start as
babashka-pod (https://github.com/rorokimdim/stash#babashka-pod-support)
T}@T{
true, false
T}@T{
false
T}
T{
\f[C]EDITOR\f[R]
T}@T{
command for editing text
T}@T{
-
T}@T{
vim
T}
.TE
.PP
Color values can be one of \f[C]black\f[R], \f[C]red\f[R],
\f[C]green\f[R], \f[C]yellow\f[R], \f[C]blue\f[R], \f[C]magenta\f[R],
\f[C]cyan\f[R], \f[C]white\f[R], \f[C]brightBlack\f[R],
\f[C]brightRed\f[R], \f[C]brightGreen\f[R], \f[C]brightYellow\f[R],
\f[C]brightBlue\f[R], \f[C]brightMagenta\f[R], \f[C]brightCyan\f[R] and
\f[C]brightWhite\f[R].
.SS Shell Autocompletion
.SS bash
.PP
\f[C]source\f[R] output from following command
.IP
.nf
\f[C]
stash --bash-completion-script \[ga]which stash\[ga]
\f[R]
.fi
.SS zsh
.PP
Create a \f[C]_stash\f[R] in your $FPATH (try
\f[C]\[ti]/.zsh/_stash\f[R]) with contents from following command
.IP
.nf
\f[C]
stash --zsh-completion-script \[ga]which stash\[ga]
\f[R]
.fi
.SS fish
.PP
\f[C]source\f[R] output from following command
.IP
.nf
\f[C]
stash --fish-completion-script \[ga]which stash\[ga]
\f[R]
.fi
.SS Credits
.IP "1." 3
Haskell (https://www.haskell.org/)
.IP "2." 3
cryptonite (https://github.com/haskell-crypto/cryptonite) and
crypto-simple (https://github.com/Risto-Stevcev/haskell-crypto-simple)
.IP "3." 3
brick (https://hackage.haskell.org/package/brick) and
optparse-applicative (https://github.com/pcapriotti/optparse-applicative)
.IP "4." 3
All of these
libraries (https://github.com/rorokimdim/stash/blob/master/package.yaml#L20)
and all the things they depend on
.IP "5." 3
Every stash file is a sqlite (https://sqlite.org/) file
.IP "6." 3
Scriping support via clojure (https://clojure.org/) and
babashka (https://github.com/borkdude/babashka)


ChangeLog.md
# Changelog for stash

## Unreleased changes


scripting-examples/python/stash.py
import atexit
import inspect
import json
import os
import subprocess
import sys
import uuid

import bcoding

POD_COMMAND = os.environ['STASH_COMMAND_FULL_PATH']
POD_PROCESS = None

STASH_FILE_PATH = "demo.stash"


class InvokeException(Exception):
    pass


def get_pod():
    """Gets pod process."""
    global POD_PROCESS
    if not POD_PROCESS:
        POD_PROCESS = subprocess.Popen(
            [POD_COMMAND],
            env=dict(BABASHKA_POD='true'),
            stdin=subprocess.PIPE,
            stderr=subprocess.PIPE,
            stdout=subprocess.PIPE)
    return POD_PROCESS


def startup():
    """Prepares to talk to stash."""
    if stash_init():
        print('✔︎ stash initialized.')
        print('Try these functions:')
        print('\n'.join(f'▸ {name}' for name, _ in get_stash_functions()))
    else:
        print(f'☠️  Invalid encryption key for "{STASH_FILE_PATH}". Failed to initialize stash.')
        print(f'\nPlease delete "{STASH_FILE_PATH}" to create one with a different key.')

    atexit.register(shutdown)


def shutdown():
    """Shuts down pod process."""
    print('Shutting down pod...')
    pod = get_pod()

    try:
        write(pod, dict(op='shutdown'))
    except ValueError as e:
        if 'closed file' in str(e):
            return
        raise

    pod.stdin.close()
    pod.terminate()
    pod.wait(timeout=0.2)

    global POD_PROCESS
    POD_PROCESS = None


def write(pod, data):
    """Writes data to pod's stdin."""
    pod.stdin.write(bcoding.bencode(data))
    pod.stdin.flush()


def read(pod):
    """Reads data from pod's stdout."""
    return bcoding.bdecode(pod.stdout)


def get_description():
    """Gets pod description."""
    pod = get_pod()

    write(pod, dict(op='describe'))
    return read(pod)


def get_stash_functions():
    """Gets a list of tuples of (stash-function-name, docstring)."""
    return sorted((name, x.__doc__) for name, x in inspect.getmembers(sys.modules[__name__])
                  if name.startswith('stash_')
                  if callable(x))


def stash_help():
    """Print some help on stash functions."""
    for name, doc in get_stash_functions():
        print(f'▸ {name}')
        print('\n'.join(line.strip() for line in doc.splitlines()))
        print()


def stash_init():
    """Initializes stash.

    The encryption key is read from STASH_ENCRYPTION_KEY environment variable.

    If `STASH_FILE_PATH` does not exist, it will be created.
    """
    return stash_invoke(
        'init',
        {'encryption-key': os.environ['STASH_ENCRYPTION_KEY'],
         'stash-path': STASH_FILE_PATH,
         'create-stash-if-missing': True})


def stash_version():
    """Gets version of stash command."""
    return stash_invoke('version')


def stash_nodes(parent_id=0):
    """Gets all nodes stored in stash.

    If a parent-node-id is provided, only nodes with that parent-id are returned.
    """
    return stash_invoke('nodes', parent_id)


def stash_tree(parent_id=0):
    """Gets all nodes stored in stash as a tree.

    Returns a dict of the form {key: [node-id, value, child-tree]}.

    If a parent-node-id is provided, only nodes with that parent-id are returned.
    """
    return stash_invoke('tree', parent_id)


def stash_tree_on_id(parent_id=0):
    """Gets all nodes stored in stash as a tree indexed by node-ids.

    Returns a dict of the form {id: {"key": key, "value": value "children": child-tree}}.

    If a parent-node-id is provided, only nodes with that parent-id are returned.
    """
    def inner(stree):
        tree = {}
        for k, (nid, value, child_stree) in stree.items():
            tree[nid] = dict(key=k, value=value, children=inner(child_stree))
        return tree

    return inner(stash_tree(parent_id=parent_id))


def stash_tree_on_id_to_paths(tree_on_id):
    """Gets paths to nodes in a tree-on-id data-structure.

    See stash_tree_on_id function.
    """
    paths = {}

    def inner(t, pid):
        for k, v in t.items():
            paths[k] = paths.get(pid, []) + [k]
            children = v['children']
            if children:
                inner(children, k)

    inner(tree_on_id, 0)
    return paths


def stash_node_versions(node_id):
    """Gets all version of a node.

    stash currently only keeps upto 10 versions.
    """
    return stash_invoke('node-versions', node_id)


def stash_get(*keys):
    """Gets value corresponding to a path of keys."""
    return stash_invoke('get', *keys)


def stash_keys(*parent_ids):
    """Gets keys under provided parent-ids.

    The root parent-id is 0.
    """
    return stash_invoke('keys', *parent_ids)


def stash_set(keys, value):
    """Sets value of a path of keys."""
    return stash_invoke('set', *(keys + [value]))


def stash_add(parent_id, key, value):
    """Adds a new node under a parent."""
    return stash_invoke('add', parent_id, key, value)


def stash_rename(node_id, new_name):
    """Renames a node."""
    return stash_invoke('rename', node_id, new_name)


def stash_update(node_id, value):
    """Updates a node's value."""
    return stash_invoke('update', node_id, value)


def stash_delete(*node_ids):
    """Deletes nodes by ids."""
    return stash_invoke('delete', *node_ids)


def stash_invoke(name, *args):
    """Invokes a stash command by name."""
    pod = get_pod()
    write(pod, dict(
        op='invoke',
        id=f'{name}-{uuid.uuid4().hex}',
        var=f'pod.rorokimdim.stash/{name}',
        args=json.dumps(args)))

    result = read(pod)

    if 'ex-message' in result:
        raise InvokeException(result['ex-message'])

    return json.loads(result['value'])


def stash_browse():
    """Launches stash terminal-ui.

    Does not currently work on windows.
    """
    subprocess.Popen([POD_COMMAND, 'browse', STASH_FILE_PATH]).wait()


startup()
